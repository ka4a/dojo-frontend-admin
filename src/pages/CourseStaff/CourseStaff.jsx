import React, {
  useCallback, useEffect, useMemo, useState,
} from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { DataTable } from '@edx/paragon';
import classNames from 'classnames';
import Button from 'react-bootstrap/esm/Button';
import { fetchCourseStaff } from '../../store/courseStaff/courseStaff.thunk';
import {
  getCourseStaffList,
  getNewStaffMember,
} from '../../store/courseStaff/courseStaff.selector';
import { setToast } from '../../store/baseLayout/baseLayout.slice';
import { AddStaffMemberModal, BaseLayout, SearchInput } from '../../components';
import {
  capitalize,
  displayOptionName,
  getOptionsFromStrings,
} from '../../utils';
import {
  requestStatus,
  roleOptions,
  statusOptions,
} from '../../utils/constants';
import {
  createActiveBadgeCell,
  createDateCell,
  createItemLinkCell,
  createTimeAgoCell,
} from '../../hocs';
import { DataTableFetcher } from '../../components/DataTableFetcher/DataTableFetcher';
import { MultiselectMenu } from '../../components/SelectMenu/MultiSelectMenu/MultiSelectMenu';
import './courseStaff.styles.scss';
import { SingleSelectMenu } from '../../components/SelectMenu/SingleSelectMenu/SingleSelectMenu';
import { fetchRoles } from '../../store/adminDashboard/adminDashboard.thunk';
import { getDashboardRoles } from '../../store/adminDashboard/adminDashboard.selector';

const STATUS_FILTER_CHOICES = ['all', 'active', 'inactive'];
const createLink = (rowData) => `/coursestaff/${rowData.user_id}`;

const COLUMNS = [
  {
    Header: 'Username',
    accessor: 'usernameAccessor',
    Cell: createItemLinkCell({
      createLink,
      isNavLink: true,
      textFieldName: 'username',
    }),
  },
  {
    Header: 'Full Name',
    accessor: 'fullNameAccessor',
    Cell: createItemLinkCell({
      createLink,
      isNavLink: true,
      textFieldName: 'full_name',
    }),
  },
  {
    Header: 'Email',
    accessor: 'email',
    disableFilters: true,
  },
  {
    Header: 'Date Join',
    accessor: 'dateJoinedTime',
    Cell: createDateCell({ fieldName: 'date_joined' }),
  },
  {
    Header: 'Last Login',
    accessor: 'loginAt',
    Cell: createTimeAgoCell({ fieldName: 'last_login' }),
  },
  {
    Header: 'Active',
    accessor: 'is_active',
    Cell: createActiveBadgeCell({ fieldName: 'is_active' }),
  },
  {
    Header: 'Staff',
    accessor: 'is_staff',
    Cell: createActiveBadgeCell({ fieldName: 'is_staff' }),
  },
];

export const CourseStaff = () => {
  const dispatch = useDispatch();

  const { data: newStaffMember } = useSelector(getNewStaffMember);
  const {
    status,
    error,
    data: courseStaffData,
    itemCount,
    pageCount,
    currentPage,
  } = useSelector(getCourseStaffList);

  const { data: dashboardRoles } = useSelector(getDashboardRoles);

  const roleChoises = useMemo(() => [roleOptions.LEARNER, ...dashboardRoles], [dashboardRoles]);

  const [searchInput, setSearchInput] = useState('');
  const [search, setSearch] = useState('');
  const [userStatus, setUserStatus] = useState(statusOptions.ALL);
  const [roles, setRoles] = useState(roleChoises);

  const filters = useMemo(() => {
    const result = {};

    if (!roles?.includes(roleOptions.LEARNER) && roles.length > 0) {
      result.role = roles.join(',');
    }

    if (search) {
      result.search = search;
    }

    if (userStatus !== statusOptions.ALL) {
      result.is_active = statusOptions[userStatus.toUpperCase()];
    }

    return result;
  }, [roles, search, userStatus]);

  const fetchData = useCallback(
    (params) => {
      dispatch(fetchCourseStaff(params));
    },
    [dispatch],
  );

  const appliedFilterNames = useMemo(() => {
    const preparedFilters = [];
    if (roles.length > 0 && roles.length < roleChoises.length) {
      preparedFilters.push(
        `Role (${roles.map((role) => displayOptionName(role)).join(', ')})`,
      );
    }

    if (userStatus !== statusOptions.ALL) {
      preparedFilters.push(`User status (${capitalize(userStatus)})`);
    }

    return preparedFilters;
  }, [userStatus, roles]);

  useEffect(() => {
    const { username, message } = newStaffMember || {};
    if (username) {
      dispatch(setToast({ body: message }));
    }
  }, [newStaffMember]);

  useEffect(() => {
    setRoles(roleChoises);
  }, [roleChoises]);

  useEffect(() => {
    if (dashboardRoles.length === 0) {
      dispatch(fetchRoles());
    }
  }, []);

  return (
    <div>
      <div
        className={classNames(
          'page-title-background',
          'd-flex',
          'align-items-center',
          'justify-content-between',
          'px-2.5',
        )}
      >
        <div className="page-title">Course Staff Management</div>
        <AddStaffMemberModal modalTitle="Add Instructor" hasEmailField />
      </div>
      <BaseLayout errors={[error]}>
        <DataTable
          manualPagination
          manualSortBy
          isLoading={status === requestStatus.LOADING}
          isPaginated
          isSortable
          itemCount={itemCount}
          pageCount={pageCount}
          initialState={{
            pageSize: courseStaffData?.length,
            pageIndex: currentPage - 1,
          }}
          data={courseStaffData}
          columns={COLUMNS}
        >
          <DataTableFetcher filters={filters} fetchData={fetchData} />
          <div
            className={classNames('course-staff__control-bar', {
              active: appliedFilterNames.length > 0,
            })}
          >
            <div className="course-staff-filters">
              <SingleSelectMenu
                variant="outline-extra"
                title="User status"
                options={STATUS_FILTER_CHOICES}
                selectedValue={userStatus}
                onSelect={setUserStatus}
              />
              <MultiselectMenu
                title="Role"
                values={roles}
                options={getOptionsFromStrings(roleChoises)}
                onChange={setRoles}
                shouldChangeOnClose
              />

              <form
                className="course-staff__search"
                onSubmit={(event) => {
                  event.preventDefault();
                  setSearch(searchInput);
                }}
              >
                <SearchInput
                  placeholder="Search user"
                  onChange={(e) => setSearchInput(e.target.value)}
                  onReset={() => {
                    setSearchInput('');
                    setSearch('');
                  }}
                />
                <Button
                  className="ml-2"
                  type="submit"
                  disabled={!searchInput.length}
                >
                  Search
                </Button>
              </form>
            </div>
            <div className="course-staff-status">
              <p>
                {appliedFilterNames.length > 0
                  ? `Filtered by: ${appliedFilterNames.join(', ')}`
                  : ''}
              </p>
              {appliedFilterNames.length > 0 && (
                <Button
                  onClick={() => {
                    setRoles(roleChoises);
                    setUserStatus(statusOptions.ALL);
                  }}
                  variant="link"
                  className="course-staff-filters__clear"
                >
                  Clear Filters
                </Button>
              )}
            </div>
          </div>
          <DataTable.Table />
          <DataTable.EmptyTable content="No results found" />
          <DataTable.TableFooter />
        </DataTable>
      </BaseLayout>
    </div>
  );
};
