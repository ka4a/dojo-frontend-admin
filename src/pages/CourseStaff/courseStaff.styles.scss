@import "@reustleco/dojo-admin-brand/paragon/variables.scss";

.course-staff {
  &__search {
    display: flex;
    margin-left: 8px;
    height: 44px;
    width: 100%;
    max-width: 528px;
  }

  &-filters {
    display: flex;
    align-items: center;
    margin-bottom: 12px;

    & > :first-child {
      margin-right: 8px;
    }

    &__clear.btn {
      padding: 0;
      font-size: 14px;
      padding: 0 4px;
    }

    .form-check-label {
      cursor: pointer;
    }
  }

  &__control-bar {
    margin: 30px 12px;

    &.active {
      margin-bottom: 12px;
    }
  }

  &__checkbox {
    pointer-events: none;
  }
}
