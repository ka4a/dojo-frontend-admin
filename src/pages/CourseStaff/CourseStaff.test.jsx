import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { AdminDashboardProvider } from '../../test/AdminDashboardProvider';
import { CourseStaff } from './CourseStaff';

let wrapper;

describe('CourseStaff', () => {
  beforeEach(() => {
    wrapper = mount(
      <MemoryRouter>
        <AdminDashboardProvider>
          <CourseStaff />
        </AdminDashboardProvider>
      </MemoryRouter>,
    );
  });

  it('should render CourseStaff component', () => {
    shallow(
      <MemoryRouter>
        <AdminDashboardProvider>
          <CourseStaff />
        </AdminDashboardProvider>
      </MemoryRouter>,
    );
  });

  it('it should render page expected HTML', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('it should render page title', () => {
    expect(
      wrapper.contains(
        <div className="page-title">Course Staff Management</div>,
      ),
    ).toEqual(true);
  });
});
