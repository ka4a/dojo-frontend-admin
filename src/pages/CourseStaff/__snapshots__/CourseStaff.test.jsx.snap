// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseStaff it should render page expected HTML 1`] = `"<div><div class=\\"page-title-background d-flex align-items-center justify-content-between px-2.5\\"><div class=\\"page-title\\">Course Staff Management</div><button type=\\"button\\" class=\\"btn btn-primary\\">Add Instructor</button></div><div class=\\"base-layout\\"><div class=\\"\\"><div class=\\"pgn__data-table-layout-wrapper\\"><div class=\\"pgn__data-table-layout-main\\"><div class=\\"pgn__data-table-wrapper\\"><div class=\\"course-staff__control-bar\\"><div class=\\"course-staff-filters\\"><div class=\\"dropdown\\"><button aria-haspopup=\\"true\\" aria-expanded=\\"false\\" type=\\"button\\" class=\\"dropdown-toggle btn btn-outline-extra\\">User status</button></div><div data-testid=\\"dropdown\\" class=\\"dropdown\\"><button aria-haspopup=\\"true\\" aria-expanded=\\"false\\" id=\\"dropdown-basic-4\\" type=\\"button\\" class=\\"dropdown-toggle btn btn-outline-extra\\">Role</button></div><form class=\\"course-staff__search\\"><div class=\\"pgn__form-group search\\"><div class=\\"pgn__form-control-decorator-group has-appended-node has-trailing-element has-floating-label\\"><input id=\\"form-field2\\" class=\\"form-control\\" value=\\"\\"><div class=\\"pgn__form-control-decorator pgn__form-control-decorator-trailing\\"><span class=\\"pgn__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\" fill=\\"currentColor\\"></path></svg></span></div><div class=\\"pgn__form-control-floating-label\\"><div class=\\"pgn__form-control-floating-label-content\\"><label class=\\"pgn__form-control-floating-label-text\\" for=\\"form-field2\\">Search user</label></div></div></div></div><button type=\\"submit\\" disabled=\\"\\" class=\\"ml-2 btn btn-primary\\">Search</button></form></div><div class=\\"course-staff-status\\"><p></p></div></div><div class=\\"pgn__data-table-container\\"><table role=\\"table\\" class=\\"pgn__data-table is-striped\\"><thead><tr role=\\"row\\"><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Username</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Full Name</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Email</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Date Join</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Last Login</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Active</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th><th colspan=\\"1\\" role=\\"columnheader\\" title=\\"Toggle SortBy\\" style=\\"cursor: pointer;\\"><span class=\\"d-flex align-items-center\\"><span>Staff</span><span class=\\"pgn__icon\\" style=\\"opacity: 0.5;\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z\\" fill=\\"currentColor\\"></path></svg></span></span></th></tr></thead><tbody role=\\"rowgroup\\"><tr role=\\"row\\" class=\\"pgn__data-table-row\\"><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><a class=\\"table-nav-link\\" href=\\"/coursestaff/32\\">test</a></td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><a class=\\"table-nav-link\\" href=\\"/coursestaff/32\\">Test</a></td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\">test@strata.co.jp</td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><time datetime=\\"1637316516000\\">November 19, 2021</time></td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><time class=\\"\\" datetime=\\"2022-05-30T11:36:05.018418Z\\" timeago-id=\\"11\\">1 week ago</time></td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><span class=\\"badge badge-success\\">True</span></td><td role=\\"cell\\" class=\\"pgn__data-table-cell-wrap\\"><span class=\\"badge badge-danger\\">False</span></td></tr></tbody></table></div><div class=\\"pgn__data-table-footer\\"><div>Showing 1 of 0.</div><nav aria-label=\\"table pagination\\" class=\\"pagination-minimal\\"><div class=\\"sr-only\\" aria-live=\\"polite\\" aria-relevant=\\"text\\" aria-atomic=\\"true\\">Page 1, Current Page, of 1</div><ul class=\\"pagination\\"><li class=\\"page-item disabled\\"><button aria-label=\\"Previous\\" class=\\"btn-icon btn-icon-primary btn-icon-md previous page-link\\" type=\\"button\\" tabindex=\\"-1\\" disabled=\\"\\"><span class=\\"btn-icon__icon-container\\"><span class=\\"pgn__icon btn-icon__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z\\" fill=\\"currentColor\\"></path></svg></span></span></button></li><li class=\\"page-item disabled\\"><button aria-label=\\"Next\\" class=\\"btn-icon btn-icon-primary btn-icon-md next page-link\\" type=\\"button\\" tabindex=\\"-1\\" disabled=\\"\\"><span class=\\"btn-icon__icon-container\\"><span class=\\"pgn__icon btn-icon__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z\\" fill=\\"currentColor\\"></path></svg></span></span></button></li></ul></nav></div></div></div></div></div></div></div>"`;
